### ISSUE TITLE
gUefiCpuPkgTokenSpaceGuid.PcdCpuApTargetCstate definition is unclear. (Bugzilla Bug 2075)

### ISSUE THREAD

**Issue by tianocore-issues** [tianocore/edk2#9775]
This issue was created automatically with [bugzilla2github](https://github.com/makubacki/bugzilla2github/tree/bz2gh_tianocore)
# Bugzilla Bug 2075
Date: 2019-08-12T16:28:58+00:00
From: @ajfish
To:   @ydong10
CC:   @ydong10, @mdkinney, @niruiyu

Last updated: 2019-08-22T20:19:56+00:00

**Comment by tianocore-issues** [comment]
## Comment 8961
Date: 2019-08-12 16:28:58 +0000
From: @ajfish

- Industry Specification: ---
- Target OS: ---
- Bugzilla Assignee(s): @ydong10

The gUefiCpuPkgTokenSpaceGuid.PcdCpuApTargetCstate .dec comments imply that this value is used for mwait programming. 

For MWAIT EAX is defined as:
[3:0] Sub C-state with a C-state. 
[7:4] Target C-state

If you look at the UefiCpuPkg MpInitLib it allways shifts the PCD Value:
  AsmMwait (CpuMpData-&gt;ApTargetCState &lt;&lt; 4, 0);

So this implies that PcdCpuApTargetCstate is defined as the MWAIT [7:4] target C state value, but the DEC seems to imply it is the MWAIT EAX value directly?

I&apos;ve got platform code from a vendor that sets PcdCpuApTargetCstate to the MWAIT value and the shift makes that end up being 0 or a C1.

The PCD also seems strange as there is no way to support sub C-states?

In conclusion either the comment is wrong in the DEC file, or the code should not be shifting left 4 bits.

**Comment by tianocore-issues** [comment]
## Comment 8964
Date: 2019-08-12 17:28:42 +0000
From: @ajfish

Also setting gUefiCpuPkgTokenSpaceGuid.PcdCpuApTargetCstate to a nono zero value in PEI with gUefiCpuPkgTokenSpaceGuid.PcdCpuApLoopMode = 2 bricked my system. 

It looked like the CPUs had not been fully setup to support C-States and the Cores not associated with the BSP crashed so hard I could not take to them with the ITP after those cores initiated the MONITOR/MWAIT.

**Comment by tianocore-issues** [comment]
## Comment 8972
Date: 2019-08-13 01:20:16 +0000
From: @ydong10

I think current implementation follow the expectation. Current code truly use PcdCpuApTargetCstate as the CState value. 

I think the current gap is not consider Sub C-state with a C-state. Maybe need another PCD for this value?

**Comment by tianocore-issues** [comment]
## Comment 9128
Date: 2019-08-21 15:57:41 +0000
From: @niruiyu

Change to Eric to follow up.

**Comment by tianocore-issues** [comment]
## Comment 9182
Date: 2019-08-22 20:19:56 +0000
From: @mdkinney

Assign to owner.  Need to determine if a code change or a comment change is required.
